File: tmb-workshop-fhl/Day5/ex5Class.R

################################################################################

setwd("tmb-workshop-fhl/Day5")
require(TMB)

################################################################################

m <- scan("ex5.dat",skip=1,n=1,quiet=T)
TT <- scan("ex5.dat",skip=3,n=m,quiet=T)
Tmax <- scan("ex5.dat",skip=5,n=1,quiet=T)
B <- matrix(scan("ex5.dat",skip=7,n=m*Tmax,quiet=T),ncol=Tmax,byrow=T)
R <- matrix(scan("ex5.dat",skip=58,n=m*Tmax,quiet=T),ncol=Tmax,byrow=T)
Phi0 <- scan("ex5.dat",skip=109,n=m,quiet=T)

data <- list(m=m,TT=TT,Tmax=Tmax,B=B,R=R,Phi0=Phi0)

################################################################################
# Hint Nproj = 0 for basic estimation
# Basic estimation
################################################################################

# Set the parameters to initial values
mu<-0.2
log_tau<- 0.1
B0<-rep(500,m)
log_sigR<-rep(0.1,m)
eps<-rep(0.5,m)

parameters <- list(dummy=0,mu=mu,log_tau=log_tau,B0=B0,
                   log_sigR=log_sigR,eps=eps)

#print(data)
#print(parameters)

compile("Ex5Class.cpp")
dyn.load(dynlib("Ex5Class"))

# test code - for checking for minimization
map<-list(mu=factor(NA),log_tau=factor(NA),
          B0=rep(factor(NA),m),log_sigR=rep(factor(NA),m),
          eps=rep(factor(NA),m))
model <- MakeADFun(data, parameters, DLL="Ex5Class",silent=T,map=map)
model$report()

xx <- model$fn(model$env$last.par)
cat(xx,model$report()$obj_fun,"\n")
#AAA

# Now estimate everything
map<-list(dummy=factor(NA))
parameters
model <- MakeADFun(data, parameters, DLL="Ex5Class",silent=T,
                   map=map, random=c("eps"))

# Bounds on the parameters - needs to be equal to the number of ESTIMATED
# parameters not equal to the length of the total number of pars in the
# parameters vectors (dummy disappears) use:
length(model$par) # how many parameters for upper and lower bounds
parameters
lowbnd= c(-Inf, # mu
          -7, # tau
          rep(0.1, m), # B0
          rep(-7, m), # log sigR
          rep(-Inf, m)) # eps

uppbnd= c(Inf, # mu
          4, # tau
          rep(1000, m), # B0
          rep(4, m),
          rep(Inf, m)) # log sig

fit <- nlminb(model$par, model$fn, model$gr, 
              control=list(rel.tol=1e-12,
                           eval.max=100000,iter.max=10000),
              lower=lowbnd,upper=uppbnd)

best <- model$env$last.par.best
print(best)
rep <- sdreport(model)
print(summary(rep))

File: tmb-workshop-fhl/Day5/Ex5Class.cpp
#include <TMB.hpp>

template <class Type> Type square(Type x){return x*x;}

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_INTEGER(m);
  DATA_IVECTOR(TT);
  DATA_INTEGER(Tmax);
  // matrix where each col is a year, each row is a stock
  DATA_MATRIX(B);
  DATA_MATRIX(R);
  DATA_VECTOR(Phi0);

  // End of data section

  PARAMETER(dummy);
  PARAMETER(mu);
  PARAMETER(log_tau);
  PARAMETER_VECTOR(B0);
  PARAMETER_VECTOR(log_sigR);
  PARAMETER_VECTOR(eps);
  // End of the estimated parameters

  vector<Type> R0(m);
  vector<Type> SigR(m);
  vector<Type> h(m);
  Type tau;
  Type beta;
  Type BH;
  Type obj_fun;
  matrix<Type> log_Rhat(m,Tmax);
  matrix<Type> log_R(m,Tmax);
  //matrix<Type> resids(m,Tmax);
  // End of the temporary variables

  // Transform the parameters
  R0 = B0/Phi0;
  tau = exp(log_tau);
  SigR = exp(log_sigR);
  
  tau = 0;
  beta = 0;
  //obj_fun = 0;
  log_Rhat.setZero();
  log_R.setZero();
  
  for (int k=0;k<m;k++) {
    for (int y=0;y<TT(k);y++) {
      log_R(k,y) = log(R(k,y));
    }
  }
  
  //resids.setZero();
  BH = 0;
  
 for (int k=0;k<m;k++) {
   // Extract beta and define h
   beta=mu+tau*eps(k);
   h(k)=(exp(beta)+Type(0.2))/(Type(1)+exp(beta));
   
   for (int y=0;y<TT(k);y++) {
     
     //Beverton-Holt model
     log_Rhat(k,y)=log(Type(4)*R0(k)*h(k)*B(k,y))-log(Phi0(k)*R0(k)*(Type(1)-h(k)) + 
       (Type(5)*h(k)-Type(1))*B(k,y)); 

     //Likelihood
     BH += pow(log_R(k,y)-log_Rhat(k,y)+(SigR(k)*SigR(k)/Type(2)),Type(2)) /
       (Type(2)*SigR(k)*SigR(k))+log(SigR(k));
     
   }
   //Adjust likelihood for random effect
   BH -= dnorm(eps(k), Type(0), Type(1), true);
   }

  //obj_fun += dummy*dummy;
 
  REPORT(beta);
  REPORT(h);
  REPORT(log_Rhat);
  ADREPORT(h);
  ADREPORT(R0);
  ADREPORT(tau);
  
  //return(obj_fun);
  return(BH);
}
